Simple product export

thisiscab opened the pull request:
This PR adds the ability to export a single product (without master-variant / variants) to Shopify. 
A previous PR was created containing the master-variant and variant export but caused problem due to extensive refactoring.

Unit tests for the exporter are not written, since I don't know what to specifically test. 
Unit tests for the base decorator were also not written, I am unsure how to actually test that either.

Suggestions are welcomed!

Let me know what you guys think of it.
Thanks!

thisiscab committed:
Include the product factory and exporter for product without variant

This will include code that will let us import products that have no
variants. The next step will be to link the creation / save of a new
product with an event that will queue the exportation to Shopify.

https://app.clubhouse.io/dynamomtl/story/33/as-a-admin-when-i-create-a-product-with-a-single-sku-it-gets-created-in-shopify-so-that-i-don-t-need-to-create-it-twice

thisiscab committed:
Export the product to Shopify when product is saved

And modify related specs to support that

thisiscab committed:
Add requests specs related to the product exportation

I wanted to make sure that we were correctly updating the product and
not creating new one when that was the case.

I have also decided to return the shopify_order instead of a true/false
response that I couldn't use.

thisiscab committed:
Add variant factory with related specs

That way we will be able to import the variant associated with the
product.

The only problem I have right now is related to the images. I'll
investigate later.

https://app.clubhouse.io/dynamomtl/story/35/as-a-admin-when-i-create-a-product-with-variants-the-master-product-and-its-variant-skus-are-created-in-shopify-so-that-i

thisiscab committed:
Modify the product factory to include the variant

thisiscab committed:
Add request specs for product with variants

Modify code to make those specs work

thisiscab committed:
Add specs related to updating an existing variant

We want to make sure that it doesn't duplicate or re-create the variant
when it is already existing on a product.

thisiscab committed:
WIP - first draft for the image import

There are tons of problems coming up. Mainly, we have a weird way of
setting and fetching the images on Glossier, I am still unsure how we'll
go around that in order to find the proper images. Right now I am using
placeholder images.

Also, in order to assign an image to a variant, the variant has to be
already saved. So we need to save the product, that saves the variant
then we assign the images to the product and when we save the product
again, it will assign the image to the variant. It is a really weird
process but I haven't found a way around that... Yet!

thisiscab committed:
Refactor the way that product factory image works

It creates a single ShopifyImage instead of an array of ShopifyImages.

I have also written specs to ensure that everything was working properly

thisiscab committed:
Save the request specs in cassettes

We don't really care about hitting the Shopify API every single time we
run those requests specs. It was taking over 40 seconds to run those
tests and now it runs almost instantly by using VCR.

Why VCR? Because this is what I am used to use and it accomplishes what
I want to do quickly.

thisiscab committed:
Use open instead of File.open

That way we can open the online URLs instead of just local files.

thisiscab committed:
Ensure that that the master variant is exported

I had totally forgotten to export the master variant of a product.
Oopsies.

thisiscab committed:
Make sure to re-create product/variant if not exist anymore

If we were to create a product on Solidus and sync it with Shopify then
delete the product from Shopify, the next time we would try to sync the
product/variant it would just crash.

Shopify would return an exception and would halt the process, in that
situation we catch the error and we re-create it to make sure that the
product exists in Shopify.

thisiscab committed:
Add more specs to make sure that everything works

Decided to temporarily remove VCR because it caused a lot of problems.
Since factory girl creates a master variant with a random sku and all
the other variants with random skus, it's hard to sync of with VCR.

I find it insane to mock each skus individually, there must be a better
way.

thisiscab committed:
Use the sku as the shopify option1 field if no option values

The option1 field in shopify is required. We are currently filling it
with option values of the solidus variant.

But the option values of a solidus variant are not required, thus we
need to use another field thus we are using the sku as the unique
identifier.

thisiscab committed:
Always use the SKU as the first option1

Shopify doesn't check for the combination of options to check if the
variant is unique, it only checks for the first option.

To ensure that it is unique per product, I have decided to always go
with the sku.

thisiscab committed:
Update travis file for specs to pass

The ENV variables were missing.

thisiscab committed:
Re-enable job the that export to shopify on product save

thisiscab committed:
Add a bunch of modifications based on CR

Mainly:
* Change factories to converters
* Better implementation for ShopifyVariant image uploads
* Better naming for certain methods
* Use prepend instead of class_eval for model decorators

thisiscab committed:
Allow the passing of a HTML renderer for the product converter

Instead of having our own method that makes it works only under a
certain scenario, we have opted to go with Redcarpet and convert the
Markdown content to HTML content. That way the content will be displayed
better than our previous ways of doing things.

Also, this let the user change the renderer to whatever rows their boat.

* This also moves the gems from the gemfile to the gemspec file.

thisiscab committed:
Add default value to the type of weight a variant has

thisiscab committed:
Fix all the specs for the product converter

There is a bunch of overhead and refactor that has to be done. There are
too many doubles and contexts everywhere.

thisiscab committed:
Fix the specs related to the variant converter

thisiscab committed:
WIP - Refactor the classes into more specialized classes

Still working on making it more readable and fixing the request specs
along the way.

thisiscab committed:
Refactor more stuff and fix more specs

thisiscab committed:
Fix the requests specs

All the requests specs are green, which means that we can continue
refactoring and create unit tests.

I have split up the requests specs in multiple different files like
suggested in the code-review.

thisiscab committed:
Change the behavior of the converter objects

This is a follow-up to what DW has suggested me

> There should already be a .update or .update_attributes method on `ShopifyAPI::Product`, as it extends ActiveResource
> https://github.com/rails/activeresource/blob/4-stable/lib/active_resource/base.rb#L1431-L1433

That way we can more easily update ShopifyAPI::Product. This breaks all
the requests specs but adds unit tests for each of those converters.

thisiscab committed:
Decorate the ActiveResource BaseDecorator

This will add the find_by and find_or_initialize_by method that will
help us further down the road.

thisiscab committed:
Remove all variant related stuff

Let's keep it simple and make one thing work at a time.

thisiscab committed:
Refactor the main product exporter class

Let's keep it as simple as possible for now.
I've also taken the liberty to remove files that were unnecessary for
this PR to pass.

thisiscab committed:
Refactor a bunch of stuff

This also adds VCR to the requests specs

I was unsure how to unit_test the exporter until receiving CR comments

cabouffard commented on app/helpers/spree/retail/presenter_helper.rb:
```diff
@@ -0,0 +1,13 @@
+    module PresenterHelper
+      def present(object, presenter_type)
+        presenter_klass = "#{presenter_type}_presenter".camelcase
+        klass = "Shopify::#{presenter_klass}".constantize
```
I'll change the namespacing to be part of the parameters

thisiscab commented on app/renderers/shopify/redcarpet_html_renderer.rb:
```diff
@@ -0,0 +1,41 @@
+    end
+
+    def render(content)
+      @converter.render(content).strip
```
Don't use an IVar here

thisiscab commented on spec/spec_helper.rb:
```diff
@@ -32,8 +34,16 @@
+end
 
 RSpec.configure do |config|
+  # WebMock.disable_net_connect!
```
Remove this

DanielWright commented:
Speaking generally, for isolated unit tests, the approach is to describe each public interface method, and exercise each of its public behaviours, by setting expectations on the commands it should send, and the results it should return. These (failing) specs will motivate your implementation: first, you'll write the simplest working version, then refactor. This is the essence of red-green-refactor.

thisiscab committed:
Update travis env variables

thisiscab committed:
Fix some minor code changes

Mostly visual stuff.

thisiscab commented:
My feeling can be resumed in: I understand how to play Hockey, but I don't know how to play Hockey.

Let's sit down once and for all and just do it.

DanielWright commented:
Are you in-office tomorrow? We can pair up to tackle this.

thisiscab commented:
I won't unfortunately. We have good internet, so SH shouldn't be a problem, even though it removes human interaction.

DanielWright commented:
OK cool. I'll schedule something in.

braidn commented on app/delegators/delegator_decorator.rb:
```diff
@@ -0,0 +1,17 @@
+  end
+end
+
+Delegator.prepend DelegatorDecorator
```
This pings my naming 'spidey sense'. Should we be more descriptive with our Decorator names as to define, top of file what exactly they are doing (aside from the decoration process)

braidn commented on app/delegators/delegator_decorator.rb:
```diff
@@ -0,0 +1,17 @@
+  end
+  alias_method :model, :__getobj__
+
+  def __setobj__(obj)
```
We aren't really modifying this method from the [Delegator class](http://ruby-doc.org/stdlib-2.3.1/libdoc/delegate/rdoc/Delegator.html) are we?

braidn commented on app/exporters/shopify/product_exporter.rb:
```diff
@@ -0,0 +1,43 @@
+    end
+
+    def perform
+      shopify_product = find_shopify_product_for(spree_product)
```
Should the following three lines be wrapped in 1 private method? Therefore the perform calls into the private methods and then has a very clear 'return' which we already have implemented.

thisiscab committed:
WIP - Unit testing

braidn commented on app/jobs/export_product_to_shopify_job.rb:
```diff
@@ -0,0 +1,10 @@
+require 'active_job'
+
+class ExportProductToShopifyJob < ActiveJob::Base
```
Love this naming scheme. It's long, but very explicit in it's function.

braidn commented on app/presenters/shopify/product_presenter.rb:
```diff
@@ -0,0 +1,17 @@
+    private
+
+    def html_renderer
+      Shopify::RedcarpetHTMLRenderer.new
```
Just wondering: how or why was Redcarpet selected? Does Redcarpet outperform Kramdown or Rdiscount?

thisiscab replied:
This is what I the most familiar with. I used Kramdown once and didn't quite like it. I have never researched about Rdiscount.

braidn commented on spec/converters/shopify/product_converter_spec.rb:
```diff
@@ -0,0 +1,29 @@
+    describe '.to_hash' do
+      subject { described_class.new(product: product).to_hash }
+
+      it { expect(subject[:title]).to eql('name') }
```
Are all of these values required? I ask because I am not a super fan of writing specs without descriptive strings.

braidn commented on spec/exporters/shopify/product_exporter_spec.rb:
```diff
@@ -0,0 +1,18 @@
+    describe '.initialize' do
+      subject { described_class.new(spree_product_id: spree_product.id, product_klass: product_klass) }
+
+      it "successfully does it's things" do
```
Successfully does it's things is a little open ended no? Maybe this could be better explained.

braidn commented on spec/requests/shopify/export_product_spec.rb:
```diff
@@ -0,0 +1,62 @@
+  it 'creates a new product' do
+    export_product!(spree_product)
+    expect(subject).to be_truthy
+    cleanup_shopify_product!(subject)
```
Can this be relegated to an after_each block?

thisiscab replied:
I tried but there are certain scenarios in the same file that are not required to be cleanup, because we clean it up in the middle of the spec thus causing spec failures.

braidn commented on spec/requests/shopify/export_product_spec.rb:
```diff
@@ -0,0 +1,62 @@
+      expect(subject).to be_truthy
+    end
+
+    after do
```
What do you think about moving this up to right after the before do block? Therefore the reader has all the context required to understand the following specs?

braidn commented on spec/requests/shopify/export_product_spec.rb:
```diff
@@ -0,0 +1,62 @@
+
+      export_product!(spree_product)
+
+      result = ShopifyAPI::Product.all.count
```
Does this read easier than:

``` ruby
expect { ShopifyAPI::Product.all.count }.to be(products_count)
```

braidn commented on spec/support/shared/shopify_helpers.rb:
```diff
@@ -0,0 +1,9 @@
+RSpec.shared_context 'shopify_helpers' do
```
Love shared contexts 👍

braidn commented:
First round CR done

thisiscab committed:
Improve code from the CR comments

Also put the unit test for the exporter in comments since this is an
ongoing process.

thisiscab commented:
Let's Re-open this PR since this will be served as the real product exportation branch.

thisiscab committed:
Update Travis config file

We are having a problem where Travis isn't fetching the proper config
variables and making the test fail. I hope this solves it.

thisiscab committed:
WIP VCR

thisiscab committed:
Add new configuration for VCR

We had trouble with Travis throwing error saying that the requests were
not found in the cassettes, not all the test passes at the moment, but
it should tell us if this way to go is the way that we should go to fix
that issue.

thisiscab committed:
Merge branch 'master' into feature/shopify/simple-product-export

thisiscab committed:
Fix some merge conflicts

thisiscab commented on spec/support/vcr.rb:
```diff
@@ -1,21 +1,21 @@
 VCR.configure do |c|
```
I am having big trouble to figure out the proper VCR config to make VCR work on TRAVIS and LOCALLY. 

Previously I had that code 

``` ruby
config.around(:each) do |example|
   if example.metadata[:type] == :request
     name = example.metadata[:full_description].split(/\s+/, 2).join('/').underscore.tr('.', '/').gsub(/[^\w\/]+/, '_').gsub(/\/$/, '')
     VCR.use_cassette(name, {}, &example)
   else
     VCR.turned_off(&example)
   end
 end
```

in the spec_helper file and all the tests were working LOCALLY but when it was hitting Travis, Travis would fail the VCR test because the request URL wouldn't be the same (why? I don't know). 

Then @DanielWright suggested to do it like we do it on Glossier. I gave it a go but now most of my VCR tests fail locally. It seems like if I re-do the same request twice in the same block, it will always retrieve the value of the first time instead of re-fetching it the second time. 

What are you guys suggestions? I have been at it for a long time now.

@braidn @DanielWright

braidn replied:
My suggestion is to throw VCR out the window and just stub the request with webmock. How many cassettes though are we creating/need to create?

thisiscab replied:
I am unsure how to count the number of cassettes since right now everything is in the same file with the "new" way of doing it. 

I would say that there are about 30 external requests being made spread out through all of our requests specs.

braidn replied:
How many of these 30 hit different endpoints on the ShopifyAPI side?

thisiscab replied:
Different endpoints maybe 10? But the params of the requests and the expected returned values are different pretty much each time.

thisiscab replied:
I have opened an issue, let's continue the discussion here:
https://github.com/glossier/solidus_retail/issues/27

thisiscab committed:
Merge branch 'master' into feature/shopify/simple-product-export

thisiscab committed:
Update cassette

thisiscab commented:
@DanielWright @braidn Test passing on travis, can I finally merge this?

thisiscab committed:
Remove empty specs file

We were working on that file in pair but never actually got really far.
